### PR TITLE
Feature: ESbuild

### PR DESCRIPTION
**Pull request by thomasdigby**
## Changes
- Update code transformation to use `esbuild` instead of `babel`, this involved a few things:
  - Replacing `@rollup/plugin-babel` & `@rollup/plugin-typescript` with `rollup-plugin-esbuild`
  - Replacing `terser` with the `esbuild` minifier
  - Replacing `babel-jest` with `es-jest` (to use esbuild under the hood)
  - Updated alias handling with `alias-hq`
  - Removing all `babel` related packages as well as `babel.config.json`
- Update `size-limit` configuration to use `preset-small-lib` to simplify the output and reduce the time to run the script, also added a few more named export tests
- Separated the creation of the types from the build script to reduce the `validate` build time (as types were unnecessary)
- Updated CI test script to remove warnings and coverage

## Results

I ran each test three times on this branch and `main` to compare and attempted to get a semi-accurate average result for each.

### Build time

| Configuration                    | Duration |
|--------------------------|----------|
| Babel & Terser                   | ~45s     |
| ESbuild                               | ~3s      |

### Test time

| Configuration                    | Duration |
|--------------------------|----------|
| Babel (babel-jest)              | ~27s     |
| ESbuild (es-jest)                | ~24s      |

### Validate time

| Configuration                    | Duration |
|--------------------------|----------|
| Babel (with types)             | ~86s     |
| ESbuild (without types)    | ~27s      |

#### ESM

| Configuration                    | Minified | Gzipped |
|--------------------------|----------|----------|
| Babel & Terser                  | 98.68 KB | 54.4 KB     |
| ESbuild                              | 83.59 KB     | 45.35 KB    |

#### CJS

| Configuration                    | Minified | Gzipped |
|--------------------------|----------|----------|
| Babel & Terser                  | 64.38 KB |  17.37 KB     |
| ESbuild                              | 59.21 KB  |  16.4 KB |

### CODE REVIEW

**Review by Mhoog — APPROVED** [review]

File: tsconfig.json [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
-    "strictNullChecks": true, // enable strict null checks as a best practice
-    "module": "ESNext", // specify module code generation
-    "jsx": "react", // use typescript to transpile jsx to js
-    "target": "esnext", // specify ECMAScript target version
-    "allowJs": true, // allow a partial TypeScript and JavaScript codebase
+    "outDir": "./dist",
+    "strictNullChecks": true,
+    "module": "ESNext",
+    "jsx": "react",
+    "target": "es2018",
```
[the remark this answers is not in the document]
Comment (hollg) [reply tsconfig.json:7]: Also curious about this 👀

**Review by hollg — APPROVED** [review]

File: package.json [hunk 2]
```diff
@@ -12,16 +12,18 @@
...
   "scripts": {
     "dev": "NODE_ENV=development rollup -c -w",
-    "build:lib": "NODE_ENV=production rollup -c && tsc --emitDeclarationOnly && tsc-alias",
+    "build:lib": "NODE_ENV=production rollup -c",
+    "build:types": "NODE_ENV=production tsc --emitDeclarationOnly && tsc-alias",
     "build:docs": "node ./scripts/generate-component-props.mjs && node ./scripts/export-documentation.mjs",
     "start:sandbox": "vite -c ./sandbox/vite.config.js",
     "add-component": "node ./scripts/add-component.mjs",
-    "prepublishOnly": "run-s clean build:lib build:docs",
+    "prepublishOnly": "run-s clean build:lib build:types build:docs",
```
[the remark this answers is not in the document]
Comment (thomasdigby) [reply package.json:20]: Just to make it clear, this change is to introduce `build:types` to the `prepublishSteps`. `build:docs` runs the Typescript/React docgen script to output the relevant prop-types and moves the MDX documentation to the `dist` directory.

File: tsconfig.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomasdigby) [reply tsconfig.json:7]: So as we were using babel previously, we were transforming the output JS into something like es2018 (babel has its own "module" support target that's similar). Prior to this change, this value was essentially ensuring that the rollup build step spat out modern `esnext` JS (with no thought about browser compatibility) before the babel step took over to transform to es2018-ish JS.

Now that we've switched to esbuild, and replaced the typescript and babel transformation steps, the ES2018 transformation now happens during the build step; this value is now relevant to what actually gets output to the dist folder. You can read more about what this value represents for esbuild here https://esbuild.github.io/content-types/#javascript

**Review by andoulla — APPROVED** [review]

File: tsconfig.json [hunk 3]
```diff
@@ -1,21 +1,21 @@
...
     "moduleResolution": "node",
+    "isolatedModules": true,
     "allowSyntheticDefaultImports": true,
     "lib": ["es6", "dom"],
     "types": ["node", "jest"],
     "baseUrl": "./src",
     "skipLibCheck": true,
     "paths": {
-      "~/*": ["./*"] // allow imports from '~/components', but not '~'
+      "~/*": ["./*"]
```
[the remark this answers is not in the document]
Comment (thomasdigby) [reply tsconfig.json:16]: I wanted to remove the comments as it makes the `tsconfig.json` invalid (comments aren't supported in JSON). As this is fairly standard behaviour for TS path aliases and the fact that we won't touch it again in future, I thought it was unnecessary to include (or repeat anywhere else in the docs)

**Pull request merged**